activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -501,6 +510,21 @@ def combine_bind_parameters(
 
       private
 
+        def with_connection_verify(exception_class)
+          yield
+        rescue exception_class
+          raise_unless_verifying_connection
+          reconnect!
+          retry
```

Comment:
This is limited to one retry, since it will only retry when `@verify_connection` is truthy and `raise_unless_verifying_connection` sets it to `@verify_connection = false`.